test/integration/job/job_test.go

```diff
@@ -1669,79 +1669,66 @@ func TestIndexedJob(t *testing.T) {
 	cases := map[string]struct {
-		podReplacementPolicyEnabled bool
-		deletePods                  bool
-		failPods                    bool
-		wantTerminating             *int32
-		wantFailed                  int
-		wantActive                  int
-		jobSpec                     *batchv1.JobSpec
+		podReplacementPolicyEnabled  bool
+		jobSpec                      *batchv1.JobSpec
+		wantActiveAfterDeletion      int
+		wantActiveAfterFailure       int
+		wantTerminatingAfterDeletion *int32
+		wantTerminatingAfterFailure  *int32
+		wantFailedAfterDeletion      int
+		wantFailedAfterFailure       int
 	}{
-		"feature flag off, delete pods and verify no terminating status": {
-			deletePods: true,
-			jobSpec:    jobSpecIndexedDefault,
-			wantActive: int(podCount),
-			wantFailed: int(podCount),
-		},
-		"feature flag true, delete pods and verify terminating status": {
-			podReplacementPolicyEnabled: true,
-			deletePods:                  true,
-			jobSpec:                     jobSpecIndexedDefault,
-			wantTerminating:             pointer.Int32(podCount),
-			wantFailed:                  int(podCount),
-		},
-		"feature flag true, delete pods, verify terminating status and recreate upon terminating": {
-			podReplacementPolicyEnabled: true,
-			deletePods:                  true,
+		"feature flag off, delete & fail pods, recreate terminating pods, and verify job status counters": {
 			jobSpec: &batchv1.JobSpec{
-				Parallelism:          pointer.Int32Ptr(podCount),
-				Completions:          pointer.Int32Ptr(podCount),
-				CompletionMode:       &indexedCompletion,
-				PodReplacementPolicy: podReplacementPolicy(batchv1.TerminatingOrFailed),
+				Parallelism:    ptr.To[int32](2),
+				Completions:    ptr.To[int32](2),
+				CompletionMode: &indexedCompletion,
 			},
-			wantTerminating: pointer.Int32(podCount),
-			wantFailed:      int(podCount),
+			wantActiveAfterDeletion: 2,
+			wantActiveAfterFailure:  2,
+			wantFailedAfterDeletion: 2,
+			wantFailedAfterFailure:  2,
 		},
-		"feature flag true, delete pods, verify terminating status and recreate once failed": {
+		"feature flag true, defaulted TerminatingOrFailed policy, delete & fail pods, recreate terminating pods, and verify job status counters": {
 			podReplacementPolicyEnabled: true,
-			deletePods:                  true,
 			jobSpec: &batchv1.JobSpec{
-				Parallelism:          pointer.Int32Ptr(podCount),
-				Completions:          pointer.Int32Ptr(podCount),
-				CompletionMode:       &indexedCompletion,
-				PodReplacementPolicy: podReplacementPolicy(batchv1.Failed),
+				Parallelism:    ptr.To[int32](2),
+				Completions:    ptr.To[int32](2),
+				CompletionMode: &indexedCompletion,
 			},
-			wantTerminating: pointer.Int32(podCount),
+			wantActiveAfterDeletion:      2,
+			wantActiveAfterFailure:       2,
+			wantFailedAfterDeletion:      2,
+			wantFailedAfterFailure:       2,
+			wantTerminatingAfterDeletion: ptr.To[int32](2),
+			wantTerminatingAfterFailure:  ptr.To[int32](2),
```

Comment:
I think there might be a bug with how terminating pods are counted in the Job status when `PodReplacementPolicy` is `TerminatingOrFailed`. The `.Status.Terminating` field never decreases to 0 unlike for `Failed` policy where it decreases to 0.
Unless I am mistaken?
cc @mimowo @alculquicondor @kannon92

Reply:
I’ll look into this tomorrow but my first thought is that this is expected.

Failed is essentially saying I want all terminating pods to be fully terminated before starting new pods.

TerminatingOrFailed says create new pods as soon as they are deleted. It takes time to create new ones but it’s pretty close to almost after deletion is registered.

so in your case what I think is happening is that you are waiting for active pods in both but in TerminatingOrFailed you are getting active pods while the terminating ones are still being deleted.

Reply:
Hmmm but I expected `Terminating` to drop to `0` after deletion & fail pods, but it didn't decrease from the value `2`.

Reply:
They do drop on `kind`. 

Again, deletion & fail for `TerminatingOrFailed` doesn't mean that it will happen immediately.  My point is that I think you would need to wait for the terminating pods to go away separately from waiting for active pods.

Reply:
We delete pods and we have this check which asserts how many pods should be terminating - https://github.com/kubernetes/kubernetes/pull/121342/files#diff-ace35385ec6f5400e8ccbfe8e9f35c2f9573d0fc05889071dc9c4b36e5e975f7R1804

Then we fail pods and do the assertion here to verify the status after faiiling them - https://github.com/kubernetes/kubernetes/pull/121342/files#diff-ace35385ec6f5400e8ccbfe8e9f35c2f9573d0fc05889071dc9c4b36e5e975f7R1815

Both checks use `validateJobsPodsStatusOnly` which does a poll and timeout is 30 seconds. I'd expect the 2 terminating pods to drop to 0 in 30 seconds.